xarray/core/groupby.py

```diff
@@ -259,11 +259,23 @@ def __init__(self, obj, group, squeeze=False, grouper=None, bins=None,
                 # TODO: sort instead of raising an error
                 raise ValueError('index must be monotonic for resampling')
             s = pd.Series(np.arange(index.size), index)
-            first_items = s.groupby(grouper).first()
-            _apply_loffset(grouper, first_items)
+            from .resample_cftime import CFTimeGrouper
+            if isinstance(grouper, CFTimeGrouper):
+                first_items = grouper.first_items(index)
+            else:
+                first_items = s.groupby(grouper).first()
+                _apply_loffset(grouper, first_items)
             full_index = first_items.index
             if first_items.isnull().any():
-                first_items = first_items.dropna()
+                if isinstance(grouper, CFTimeGrouper):
+                    index_dict = dict(zip(np.arange(first_items.size),
+                                          first_items.index.values))
+                    first_items.index = np.arange(first_items.size)
+                    first_items = first_items.dropna()
+                    first_items.index = [index_dict[i] for i in
+                                         first_items.index.values]
+                else:
+                    first_items = first_items.dropna()
```

Comment:
I would suggest moving all this logic into a dedicate helper function that returns `full_index` and `first_items` -- possibly with entirely separate paths for CFTimeGrouper / pandas.Grouper. That would help keep the down complexity in `GroupBy.__init__`.

Reply:
Is this what you meant:

```python
        if grouper is not None:
            index = safe_cast_to_index(group)
            if not index.is_monotonic:
                # TODO: sort instead of raising an error
                raise ValueError('index must be monotonic for resampling')
            s = pd.Series(np.arange(index.size), index)
            full_index, first_items = self._get_index_and_items(
                index, s, grouper)
```

```python
    @staticmethod
    def _get_index_and_items(index, s, grouper):
        from .resample_cftime import CFTimeGrouper
        if isinstance(grouper, CFTimeGrouper):
            first_items = grouper.first_items(index)
            full_index = first_items.index
            if first_items.isnull().any():
                index_dict = dict(zip(np.arange(first_items.size),
                                      first_items.index.values))
                first_items.index = np.arange(first_items.size)
                first_items = first_items.dropna()
                first_items.index = [index_dict[i] for i in
                                     first_items.index.values]
        else:
            first_items = s.groupby(grouper).first()
            _apply_loffset(grouper, first_items)
            full_index = first_items.index
            if first_items.isnull().any():
                first_items = first_items.dropna()
        return full_index, first_items
```

Reply:
@jwenfai Yes, something like that looks great! But please it make just a function -- usually static methods are unnecessary and a little confusing.

Reply:
Pushed the changes, travis-ci build is failing and I can't diagnose what's wrong from the error message (something to do with conda it seems).